File: src/main/java/com/temporal/agentic/workflows/NonAgenticDagParentWorkflowImpl.java
package com.temporal.agentic.workflows;

import com.temporal.agentic.engine.*;
import com.temporal.agentic.models.*;
import io.temporal.workflow.*;
import org.slf4j.Logger;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.CompletableFuture;

public class NonAgenticDagParentWorkflowImpl implements NonAgenticDagParentWorkflow {

    private static final Logger logger = Workflow.getLogger(NonAgenticDagParentWorkflowImpl.class);

    private String workflowStatus = "STARTED";
    private WorkflowContext context;
    private Map<String, String> nodeStatuses = new TreeMap<>();
    private DagSpec dagSpec;
    private DagExecutor dagExecutor;
    private ContextMergeEngine mergeEngine;

    @Override
    public FinalResponse executeWorkflow(InputRequest request, DagSpec dagSpec) {
        this.dagSpec = dagSpec;
        this.dagExecutor = new DagExecutor();
        this.mergeEngine = dagExecutor.getMergeEngine();

        // Initialize context
        context = new WorkflowContext();
        context.getTraceContext().put("workflowRunId", Workflow.getInfo().getWorkflowId());
        context.getTraceContext().put("requestId", request.getRequestId());

        Map<String, Object> initialData = new TreeMap<>();
        if (request.getData() != null) {
            initialData.putAll(request.getData());
        }
        context.setData(initialData);

        // Initialize node statuses
        for (NodeDefinition node : dagSpec.getNodes()) {
            nodeStatuses.put(node.getId(), "PENDING");
        }

        logger.info("Starting non-agentic workflow for request: {}", request.getRequestId());

        try {
            executeDAG();
            workflowStatus = "COMPLETED";
            return buildFinalResponse(request, "SUCCESS");
        } catch (Exception e) {
            logger.error("Workflow execution failed", e);
            workflowStatus = "FAILED";
            return buildFinalResponse(request, "FAILED");
        }
    }

    private void executeDAG() {
        Map<String, NodeDefinition> nodeMap = dagExecutor.buildNodeMap(dagSpec);
        List<String> executedNodes = new ArrayList<>();

        while (true) {
            DagExecutor.DagExecutionPlan plan = dagExecutor.planExecution(dagSpec, context);
            if (plan.getStages().isEmpty()) {
                break;
            }

            for (DagExecutor.DagExecutionPlan.Stage stage : plan.getStages()) {
                List<CompletableFuture<NodeExecutionResponse>> futures = new ArrayList<>();

                for (String nodeId : stage.getNodeIds()) {
                    nodeStatuses.put(nodeId, "RUNNING");
                    futures.add(executeNodeAsync(nodeId, nodeMap.get(nodeId)));
                }

                // Wait for all nodes in stage to complete
                Workflow.allOf(futures).get();

                // Collect and merge responses
                for (CompletableFuture<NodeExecutionResponse> future : futures) {
                    try {
                        NodeExecutionResponse response = future.get();
                        executedNodes.add(response.getNodeId());
                        mergeContextFromResponse(response);
                    } catch (Exception e) {
                        logger.error("Failed to get node response", e);
                    }
                }
            }
        }

        logger.info("DAG execution completed. Executed nodes: {}", executedNodes);
    }

    private CompletableFuture<NodeExecutionResponse> executeNodeAsync(String nodeId, NodeDefinition nodeDef) {
        return Workflow.async(() -> {
            try {
                NodeChildWorkflow child = Workflow.newChildWorkflowStub(NodeChildWorkflow.class,
                        ChildWorkflowOptions.newBuilder()
                                .setTaskQueue(nodeDef.getTaskQueue() != null ? nodeDef.getTaskQueue() : "default")
                                .setStartToCloseTimeout(Duration.ofSeconds(600))
                                .setParentClosePolicy(ParentClosePolicy.PARENT_CLOSE_POLICY_REQUEST_CANCEL)
                                .build());

                ContextSnapshot snapshot = mergeEngine.captureContextSnapshot(context);
                Map<String, Object> nodeInputs = resolveNodeInputs(nodeDef, nodeMap(), context);

                NodeExecutionRequest childRequest = new NodeExecutionRequest(
                        Workflow.getInfo().getWorkflowId(),
                        nodeId,
                        nodeDef.getType(),
                        nodeInputs,
                        snapshot,
                        nodeId + "-" + Workflow.getInfo().getWorkflowId(),
                        1,
                        context.getTraceContext()
                );

                NodeExecutionResponse response = child.executeNode(childRequest);
                nodeStatuses.put(nodeId, response.getStatus());

                logger.info("Node {} completed with status: {}", nodeId, response.getStatus());
                return response;
            } catch (Exception e) {
                logger.error("Node {} execution error: {}", nodeId, e.getMessage());
                NodeExecutionResponse errorResponse = new NodeExecutionResponse();
                errorResponse.setNodeId(nodeId);
                errorResponse.setStatus("FAILED");
                errorResponse.setErrors(new ArrayList<>(List.of(e.getMessage())));
                return errorResponse;
            }
        });
    }

    private Map<String, NodeDefinition> nodeMap() {
        return dagExecutor.buildNodeMap(dagSpec);
    }

    private void mergeContextFromResponse(NodeExecutionResponse response) {
        if ("SUCCESS".equals(response.getStatus()) || "SKIPPED".equals(response.getStatus())) {
            mergeEngine.mergeContextDelta(context, response, System.currentTimeMillis());
        }
    }

    private Map<String, Object> resolveNodeInputs(NodeDefinition nodeDef, Map<String, NodeDefinition> nodeMap,
                                                  WorkflowContext context) {
        Map<String, Object> resolved = new TreeMap<>();
        if (nodeDef.getInputs() == null) {
            return resolved;
        }

        for (Map.Entry<String, Object> entry : nodeDef.getInputs().entrySet()) {
            Object value = entry.getValue();
            if (value instanceof String && ((String) value).startsWith("ref(")) {
                String refPath = ((String) value).substring(4, ((String) value).length() - 1);
                resolved.put(entry.getKey(), resolveContextPath(refPath, context));
            } else if ("required".equals(value)) {
                // Will be filled from input request
                resolved.put(entry.getKey(), null);
            } else {
                resolved.put(entry.getKey(), value);
            }
        }

        return resolved;
    }

    private Object resolveContextPath(String path, WorkflowContext context) {
        String[] parts = path.split("\\.");
        Object current = context.getData().getOrDefault("nodes", new TreeMap<>());

        for (String part : parts) {
            if (current instanceof Map) {
                current = ((Map<?, ?>) current).get(part);
            } else {
                return null;
            }
        }
        return current;
    }

    private FinalResponse buildFinalResponse(InputRequest request, String status) {
        FinalResponse response = new FinalResponse();
        response.setRequestId(request.getRequestId());
        response.setStatus(status);

        WorkflowStatus workflowStat = new WorkflowStatus();
        workflowStat.setStatus(workflowStatus);
        workflowStat.setTotalNodes(dagSpec.getNodes().size());
        long completed = nodeStatuses.values().stream().filter(s -> "SUCCESS".equals(s) || "SKIPPED".equals(s)).count();
        workflowStat.setCompletedNodes((int) completed);
        long failed = nodeStatuses.values().stream().filter(s -> "FAILED".equals(s)).count();
        workflowStat.setFailedNodes((int) failed);

        response.setWorkflowStatus(workflowStat);
        response.setOutputs(new TreeMap<>(context.getData()));
        response.setContextSummary(new TreeMap<>());
        response.getContextSummary().put("version", context.getVersion());
        response.getContextSummary().put("nodeCount", nodeStatuses.size());
        response.setErrors(new ArrayList<>());

        return response;
    }

    @Override
    public String getStatus() {
        return workflowStatus;
    }

    @Override
    public Map<String, String> getNodeStatuses() {
        return new TreeMap<>(nodeStatuses);
    }

    @Override
    public Map<String, Object> getContextSummary() {
        Map<String, Object> summary = new TreeMap<>();
        summary.put("contextVersion", context.getVersion());
        summary.put("nodeStates", context.getNodeStates().keySet());
        summary.put("status", workflowStatus);
        return summary;
    }
}

File: src/main/kotlin/com/temporal/agentic/workflows/NonAgenticDagParentWorkflowImpl.kt
package com.temporal.agentic.workflows

import com.temporal.agentic.engine.ContextMergeEngine
import com.temporal.agentic.engine.DagExecutor
import com.temporal.agentic.models.*
import io.temporal.workflow.*
import org.slf4j.Logger
import java.time.Duration
import java.util.*

class NonAgenticDagParentWorkflowImpl : NonAgenticDagParentWorkflow {
    
    private val logger: Logger = Workflow.getLogger(NonAgenticDagParentWorkflowImpl::class.java)
    
    private var workflowStatus = "STARTED"
    private lateinit var context: WorkflowContext
    private val nodeStatuses = TreeMap<String, String>()
    private lateinit var dagSpec: DagSpec
    private lateinit var dagExecutor: DagExecutor
    private lateinit var mergeEngine: ContextMergeEngine
    
    override fun executeWorkflow(request: InputRequest, dagSpec: DagSpec): FinalResponse {
        this.dagSpec = dagSpec
        this.dagExecutor = DagExecutor()
        this.mergeEngine = dagExecutor.getMergeEngine()
        
        context = WorkflowContext()
        context.traceContext["workflowRunId"] = Workflow.getInfo().workflowId
        context.traceContext["requestId"] = request.requestId!!
        
        val initialData = TreeMap<String, Any>()
        request.data?.let { initialData.putAll(it) }
        context.data = initialData
        
        for (node in dagSpec.nodes) {
            nodeStatuses[node.id!!] = "PENDING"
        }
        
        logger.info("Starting non-agentic workflow for request: {}", request.requestId)
        
        return try {
            executeDAG()
            workflowStatus = "COMPLETED"
            buildFinalResponse(request, "SUCCESS")
        } catch (e: Exception) {
            logger.error("Workflow execution failed", e)
            workflowStatus = "FAILED"
            buildFinalResponse(request, "FAILED")
        }
    }
    
    private fun executeDAG() {
        val nodeMap = dagExecutor.buildNodeMap(dagSpec)
        val executedNodes = mutableListOf<String>()
        
        while (true) {
            val plan = dagExecutor.planExecution(dagSpec, context)
            if (plan.stages.isEmpty()) {
                break
            }
            
            for (stage in plan.stages) {
                val futures = mutableListOf<Promise<NodeExecutionResponse>>()
                
                for (nodeId in stage.nodeIds) {
                    nodeStatuses[nodeId] = "RUNNING"
                    futures.add(executeNodeAsync(nodeId, nodeMap[nodeId]!!))
                }
                
                Promise.allOf(futures).get()
                
                for (future in futures) {
                    try {
                        val response = future.get()
                        executedNodes.add(response.nodeId!!)
                        mergeContextFromResponse(response)
                    } catch (e: Exception) {
                        logger.error("Failed to get node response", e)
                    }
                }
            }
        }
        
        logger.info("DAG execution completed. Executed nodes: {}", executedNodes)
    }
    
    private fun executeNodeAsync(nodeId: String, nodeDef: NodeDefinition): Promise<NodeExecutionResponse> {
        return Async.function {
            try {
                val child = Workflow.newChildWorkflowStub(
                    NodeChildWorkflow::class.java,
                    ChildWorkflowOptions.newBuilder()
                        .setTaskQueue(nodeDef.taskQueue ?: "default")
                        .setWorkflowExecutionTimeout(Duration.ofSeconds(600))
                        .build()
                )
                
                val snapshot = mergeEngine.captureContextSnapshot(context)
                val nodeInputs = resolveNodeInputs(nodeDef, nodeMap(), context)
                
                val childRequest = NodeExecutionRequest(
                    workflowRunId = Workflow.getInfo().workflowId,
                    nodeId = nodeId,
                    nodeType = nodeDef.type,
                    nodeInput = nodeInputs,
                    contextSnapshot = snapshot,
                    idempotencyKey = "$nodeId-${Workflow.getInfo().workflowId}",
                    attempt = 1,
                    traceContext = context.traceContext
                )
                
                val response = child.executeNode(childRequest)
                nodeStatuses[nodeId] = response.status!!
                
                logger.info("Node {} completed with status: {}", nodeId, response.status)
                response
            } catch (e: Exception) {
                logger.error("Node {} execution error: {}", nodeId, e.message)
                NodeExecutionResponse(
                    nodeId = nodeId,
                    status = "FAILED",
                    errors = listOf(e.message ?: "Unknown error")
                )
            }
        }
    }
    
    private fun nodeMap(): Map<String, NodeDefinition> {
        return dagExecutor.buildNodeMap(dagSpec)
    }
    
    private fun mergeContextFromResponse(response: NodeExecutionResponse) {
        if (response.status == "SUCCESS" || response.status == "SKIPPED") {
            mergeEngine.mergeContextDelta(context, response, System.currentTimeMillis())
        }
    }
    
    private fun resolveNodeInputs(
        nodeDef: NodeDefinition,
        nodeMap: Map<String, NodeDefinition>,
        context: WorkflowContext
    ): Map<String, Any> {
        val resolved = TreeMap<String, Any>()
        nodeDef.inputs ?: return resolved
        
        for ((key, value) in nodeDef.inputs!!) {
            when {
                value is String && value.startsWith("ref(") -> {
                    val refPath = value.substring(4, value.length - 1)
                    resolveContextPath(refPath, context)?.let { resolved[key] = it }
                }
                value == "required" -> {
                    // Will be filled from input request
                }
                else -> {
                    resolved[key] = value
                }
            }
        }
        
        return resolved
    }
    
    private fun resolveContextPath(path: String, context: WorkflowContext): Any? {
        val parts = path.split(".")
        var current: Any? = context.data["nodes"] ?: TreeMap<String, Any>()
        
        for (part in parts) {
            current = when (current) {
                is Map<*, *> -> current[part]
                else -> return null
            }
        }
        
        return current
    }
    
    private fun buildFinalResponse(request: InputRequest, status: String): FinalResponse {
        val workflowStat = WorkflowStatus(
            status = workflowStatus,
            totalNodes = dagSpec.nodes.size,
            completedNodes = nodeStatuses.values.count { it == "SUCCESS" || it == "SKIPPED" },
            failedNodes = nodeStatuses.values.count { it == "FAILED" }
        )
        
        val contextSummary = TreeMap<String, Any>()
        contextSummary["version"] = context.version
        contextSummary["nodeCount"] = nodeStatuses.size
        
        return FinalResponse(
            requestId = request.requestId,
            status = status,
            workflowStatus = workflowStat,
            outputs = TreeMap(context.data),
            contextSummary = contextSummary,
            errors = emptyList()
        )
    }
    
    override fun getStatus(): String = workflowStatus
    
    override fun getNodeStatuses(): Map<String, String> = TreeMap(nodeStatuses)
    
    override fun getContextSummary(): Map<String, Any> {
        return TreeMap<String, Any>().apply {
            put("contextVersion", context.version)
            put("nodeStates", context.nodeStates.keys)
            put("status", workflowStatus)
        }
    }
}
